presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java

```diff
@@ -2465,9 +2578,29 @@ public void revokeTablePrivileges(ConnectorSession session, SchemaTableName sche
     }
 
     @Override
-    public List<GrantInfo> listTablePrivileges(ConnectorSession session, SchemaTablePrefix schemaTablePrefix)
+    public Map<SchemaTableName, Optional<List<GrantInfo>>> listTablePrivilegesWithRedirection(ConnectorSession session, SchemaTablePrefix schemaTablePrefix)
     {
-        return accessControlMetadata.listTablePrivileges(session, listTables(session, schemaTablePrefix));
+        Map<SchemaTableName, Optional<ImmutableList.Builder<GrantInfo>>> tablePrivileges = new HashMap<>();
+
+        List<SchemaTableName> tables = listTables(session, schemaTablePrefix);
+        if (isRedirectToIcebergEnabled(session)) {
+            ImmutableList.Builder<SchemaTableName> nonIcebergTables = ImmutableList.builder();
+            for (SchemaTableName table : tables) {
+                if (isExistingIcebergTable(session, table)) {
+                    tablePrivileges.put(table, Optional.empty());
+                }
+                else {
+                    nonIcebergTables.add(table);
+                }
+            }
+            tables = nonIcebergTables.build();
+        }
+        accessControlMetadata.listTablePrivileges(session, tables).forEach(grantInfo -> {
+            SchemaTableName tableName = grantInfo.getSchemaTableName();
+            tablePrivileges.putIfAbsent(tableName, Optional.of(ImmutableList.builder()));
+            tablePrivileges.get(tableName).get().add(grantInfo);
+        });
+        return tablePrivileges.entrySet().stream().collect(toImmutableMap(Entry::getKey, entry -> entry.getValue().isPresent() ? Optional.of(entry.getValue().get().build()) : Optional.empty()));
```

Comment:
It seems the implementation of `listTablePrivilegesWithRedirection` method could be simplified.
Before suggesting that i have a question on the `listTablePrivilegesWithRedirection` return type. Added there.